File: 0957-minimum-add-to-make-parentheses-valid/solution.java
class Solution {
    public int minAddToMakeValid(String s) {
        int cnt = 0;
        Stack<Character> stk = new Stack<Character>();
        for(int i=0; i<s.length(); i++){
            char ch = s.charAt(i);
            if(ch == ')' ){
                if(stk.size() >= 1){
                    stk.pop();
                }else{
                    cnt++;
                }
            }
            else{  // definitly it is '('
                stk.push(ch);
            }
        }
        cnt += stk.size();
        return cnt;
    }
}

File: 0957-minimum-add-to-make-parentheses-valid/solution.cpp
class Solution {
public:
    int minAddToMakeValid(string s) {
        stack<char> open;
        stack<char> close;
        for(char ch : s){
            if( ch == '(') open.push('(');
            else{
                if(open.size() == 0){
                    close.push(')');
                }else open.pop();
            }
        }
        return (open.size() + close.size());
    }
};
